### PR TITLE
BAH-4351 | Migrated NumericBox, Date and DateTime components

### PR DESCRIPTION
**Pull request by sahupadma**
JIRA - https://bahmni.atlassian.net/browse/BAH-4351

Currently We have done migration for NumericBox, Date and DateTime components to bahmni-design-system Carbon Component. and added CSS as per required

### CODE REVIEW

**Review by ravinderkabli — COMMENTED** [review]
## Review for BAH-4351

Good progress on the Carbon migration for Date, DateTime, and NumericBox. The components generally follow the established TextBox pattern well. Found a few critical issues that need fixing before merge — primarily the DateTime key casing in CarbonContainer (making it unreachable), the `disabled` prop mismatch in NumericBox, and missing absolute range validation props.

Found **3 blocking** issue(s), **5 suggestion(s)**, and **4 nit(s)**.

**Acceptance Criteria**: ~7/13 met

| Criterion | Status |
|-----------|--------|
| Components importable | ⚠️ DateTime unreachable via CarbonContainer (key casing) |
| Design system usage | ⚠️ Missing `invalid` prop on Date/DateTime Carbon inputs |
| Variants/sizes | 🔵 Hardcoded — legacy also lacks this |
| Disabled state | ⚠️ NumericBox uses wrong prop (`disabled` vs `enabled`) |
| Error handling | ⚠️ `touched` gate blocks form-level validation |
| onChange fires | ✅ |
| onBlur fires | 🔵 Legacy also doesn't have this |
| Controlled component | ⚠️ NumericBox coerces undefined→0 |
| Value extraction | ✅ |
| Data serialization | ✅ |
| Data deserialization | ⚠️ DateTime stale state bug |
| Controlled update | ✅ |

### What's Done Well
- Comprehensive NumericBox test suite (611 lines)
- Consistent use of `Validator.getErrors()` shared helper
- Clean Carbon component integration (DatePicker, TimePicker, NumberInput)
- Good normal range display with edge case handling (zero values)
- Well-documented SCSS overrides explaining why `!important` is needed

🤖 Generated with [Claude Code](https://claude.ai/code)

File: src/components/bahmni-design-system/CarbonContainer.jsx [hunk 1]
```diff
@@ -3,10 +3,16 @@ import { Container } from 'components/Container.jsx';
...
+import { Date } from 'components/bahmni-design-system/Date';
+import { DateTime } from 'components/bahmni-design-system/DateTime';
+import { NumericBox } from 'components/bahmni-design-system/NumericBox';
 
 const carbonComponents = {
   text: TextBox,
   obsgroupcontrol: ObsGroupControl,
+  date: Date,
+  dateTime: DateTime,
+  numeric: NumericBox,
```
Comment (ravinderkabli) [inline src/components/bahmni-design-system/CarbonContainer.jsx:15]: blocking: This key will never match. `carbonStore.getRegisteredComponent(type)` calls `type.toLowerCase()` before lookup, so `'dateTime'` becomes `'datetime'` — which doesn't match the camelCase key `dateTime`. The Carbon DateTime component is silently unreachable; the legacy one is served instead.

Change to lowercase:
```js
datetime: DateTime,
```
This matches how all other keys (`text`, `obsgroupcontrol`, `date`, `numeric`) are already lowercase.

File: src/components/bahmni-design-system/NumericBox.jsx [hunk 2]
```diff
@@ -0,0 +1,166 @@
+import { NumberInput } from '@bahmni/design-system';
+import { Validator } from 'src/helpers/Validator';
+import isEmpty from 'lodash/isEmpty';
+import constants from 'src/constants';
+import React, { useState, useEffect } from 'react';
+
+export const NumericBox = ({
+  value,
+  onChange,
+  disabled,
```
Comment (ravinderkabli) [inline src/components/bahmni-design-system/NumericBox.jsx]: blocking: `Container.jsx` computes `enabled: !readonly` and `ObsControl` forwards `enabled` to all components. `TextBox`, `Date`, and `DateTime` all accept `enabled` and derive `disabled={!enabled}`. Since NumericBox destructures `disabled` and `readOnly` directly, both will always be `undefined` in the real form pipeline — the input stays permanently interactive.

Suggested fix — replace with:
```js
enabled,
```
And in the render: `disabled={!enabled}`.
Also add `NumericBox.defaultProps = { enabled: true }` to match TextBox.

File: src/components/bahmni-design-system/NumericBox.jsx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  const [hasWarnings, setHasWarnings] = useState(false);
+  const [isInitialized, setIsInitialized] = useState(false);
+
+  const _getErrors = (val) => {
+    const allValidations = [
+      constants.validations.allowRange,
+      constants.validations.minMaxRange,
+      ...validations,
+    ];
+    const params = {
```
Comment (ravinderkabli) [inline src/components/bahmni-design-system/NumericBox.jsx:31]: blocking: The legacy NumericBox passes `minAbsolute`/`maxAbsolute` from `lowAbsolute`/`hiAbsolute` props. This params object only has `minNormal`/`maxNormal`, so `Validator.minMaxRange` always sees `undefined` for the absolute bounds — absolute range validation is silently broken.

Add the missing props:
```js
const params = {
  minNormal: lowNormal,
  maxNormal: hiNormal,
  minAbsolute: lowAbsolute,
  maxAbsolute: hiAbsolute,
};
```
And destructure `lowAbsolute, hiAbsolute` from the component props.

File: test/components/bahmni-design-system/Date.test.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+describe('Date', () => {
+  const mockOnChange = jest.fn();
+
+  beforeEach(() => {
+    mockOnChange.mockClear();
+  });
+
+  test('renders Date with value prop', () => {
+    render(<Date value="test" onChange={mockOnChange} />);
+    // Add assertions
```
Comment (ravinderkabli) [inline test/components/bahmni-design-system/Date.test.js]: suggestion: All 6 tests in this file (and all 6 in DateTime.test.js) have empty bodies — they pass vacuously with zero coverage. Also, required props (`formFieldPath`, `validate`, `validateForm`, `validations`) are missing, and some tests pass `disabled`/`readOnly`/`hasErrors` which aren't props on these components (they use `enabled` and track errors internally).

Please implement assertions — at minimum: render with all required props, assert onChange on mount, assert error class on validation failure, assert disabled via `enabled={false}`.

File: src/components/bahmni-design-system/Date.jsx [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+  _getErrors(value) {
+    const validations = this.props.validations;
+    const controlDetails = { validations, value };
+    return Validator.getErrors(controlDetails);
+  }
+
+  render() {
+    const { conceptUuid, label, enabled } = this.props;
+    const displayHasErrors = this.state.touched && this.state.hasErrors;
```
Comment (ravinderkabli) [inline src/components/bahmni-design-system/Date.jsx]: suggestion: The `touched` gate deviates from both the legacy Date component and the Carbon TextBox — neither uses `touched`. When the form triggers validation (`validate=true`), `hasErrors` updates but `touched` stays `false`, so error styling never appears until the user interacts.

Consider removing the `touched` gate to match TextBox:
```js
const displayHasErrors = this.state.hasErrors;
```
The same applies to DateTime (line 141).

File: src/components/bahmni-design-system/NumericBox.jsx [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+      hiNormal !== null &&
+      (!lowNormal && lowNormal !== 0)
+    ) {
+      return `(< ${hiNormal})`;
+    }
+    return null;
+  };
+
+  // Ensure value is always numeric (never string) for Carbon NumberInput.toFixed() compatibility
+  const getNumericValue = () => {
```
Comment (ravinderkabli) [inline src/components/bahmni-design-system/NumericBox.jsx:143]: suggestion: `getNumericValue()` returns `0` for undefined/null/empty. The legacy NumericBox starts with an empty input. This is a UX regression — a field that's never been filled shows `0`, making it impossible to distinguish "no value" from "the value is zero."

Consider using `allowEmpty` on Carbon's NumberInput if available, or returning `''` for empty states.

File: src/components/bahmni-design-system/NumericBox.jsx [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+    // Don't show errors on mount - only show them after user interaction or form validation
+    setHasErrors(false);
+    setHasWarnings(false);
+
+    // Call onChange on mount with the value if it exists
+    if (typeof value !== 'undefined' || validateForm) {
+      onChange({
+        value,
+        errors: [],
```
Comment (ravinderkabli) [inline src/components/bahmni-design-system/NumericBox.jsx:61]: suggestion: Hardcoded empty errors array on mount. The legacy NumericBox and the other new components (Date, DateTime) compute real errors via `_getErrors(value)` on mount. When `validateForm=true` with a mandatory field, the parent receives zero errors — skipping initial validation.

Replace with:
```js
errors: _getErrors(value),
```

File: src/components/bahmni-design-system/DateTime.jsx [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+    if (this.props.validate !== prevProps.validate) {
+      const errors = this._getAllErrors(this.state.dateValue, this.state.timeValue);
+      const hasErrors = this._hasErrors(errors);
+      if (this.state.hasErrors !== hasErrors) {
+        this.setState({ hasErrors });
+      }
+    }
+
+    const errors = this._getAllErrors(this.state.dateValue, this.state.timeValue);
```
Comment (ravinderkabli) [inline src/components/bahmni-design-system/DateTime.jsx]: suggestion: After `setState({ dateValue, timeValue })` on line 48, `this.state` still holds the OLD values (React batches state updates). This line reads stale `this.state.dateValue`/`this.state.timeValue`, so `onChange` may fire with incorrect error state.

Use the locally computed values instead:
```js
const { dateValue: currentDate, timeValue: currentTime } = 
  !isEqual(this.props.value, prevProps.value) 
    ? this._parseValue(this.props.value) 
    : this.state;
const errors = this._getAllErrors(currentDate, currentTime);
```

File: src/components/bahmni-design-system/Date.jsx [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+    return (
+      <div className={classNames({
+        'form-builder-error': displayHasErrors,
+        'form-builder-warning': false
+      })}>
+        <DatePicker
```
Comment (ravinderkabli) [inline src/components/bahmni-design-system/Date.jsx:127]: nit: TextBox passes `invalid={this.state.hasErrors}` to the Carbon TextArea. Date and DateTime wrap in a div with CSS class instead. Consider passing `invalid` to `DatePickerInput` for consistent Carbon error styling across components.

File: src/components/bahmni-design-system/Date.jsx [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+  conceptUuid: PropTypes.string,
+  enabled: PropTypes.bool,
+  formFieldPath: PropTypes.string.isRequired,
+  label: PropTypes.string,
+  onChange: PropTypes.func.isRequired,
+  validate: PropTypes.bool.isRequired,
+  validateForm: PropTypes.bool.isRequired,
+  validations: PropTypes.array.isRequired,
+  value: PropTypes.string,
+};
```
Comment (ravinderkabli) [inline src/components/bahmni-design-system/Date.jsx:158]: nit: TextBox declares `defaultProps = { enabled: true }`. Date and DateTime omit this. It works by accident (`!undefined === true`), but adding `Date.defaultProps = { enabled: true }` would be consistent with the established pattern. Same for DateTime.

File: src/components/bahmni-design-system/Date.jsx [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+  _formatDate(date) {
+    if (!date) return undefined;
+    const year = date.getFullYear();
+    const month = String(date.getMonth() + 1).padStart(2, '0');
+    const day = String(date.getDate()).padStart(2, '0');
+    return `${year}-${month}-${day}`;
+  }
+
+  _parseDate(dateString) {
```
Comment (ravinderkabli) [inline src/components/bahmni-design-system/Date.jsx]: nit: `_parseDate` is never called (dead code). Also, since this class is named `Date`, `new Date(dateString)` on line 96 refers to the component class itself — not the global Date constructor. If ever called, it would create a component instance instead of a date object. Recommend removing this method.

File: src/components/bahmni-design-system/Date.jsx [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  componentDidMount() {
+    const { value, validateForm } = this.props;
+    if (this.state.hasErrors || typeof value !== 'undefined' || validateForm) {
+      this.props.onChange({ value, errors: this._getErrors(value), triggerControlEvent: false, calledOnMount: true });
+    }
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
```
Comment (ravinderkabli) [inline src/components/bahmni-design-system/Date.jsx:25]: nit: `shouldComponentUpdate` checks `hasErrors` but not `touched`. If `handleChange` sets `touched: true` while `hasErrors` doesn't change, the component skips re-rendering. This becomes moot if the `touched` gate is removed per the suggestion on line 125.

File: src/components/bahmni-design-system/Date.jsx (hunk 11 above)
Comment (sahupadma) [reply src/components/bahmni-design-system/Date.jsx]: removed

File: src/components/bahmni-design-system/Date.jsx (hunk 10 above)
Comment (sahupadma) [reply src/components/bahmni-design-system/Date.jsx:158]: added

File: src/components/bahmni-design-system/Date.jsx (hunk 9 above)
Comment (sahupadma) [reply src/components/bahmni-design-system/Date.jsx:127]: added invalid to DatePickerInput

File: src/components/bahmni-design-system/NumericBox.jsx (hunk 7 above)
Comment (sahupadma) [reply src/components/bahmni-design-system/NumericBox.jsx:61]: done

File: src/components/bahmni-design-system/Date.jsx (hunk 5 above)
Comment (sahupadma) [reply src/components/bahmni-design-system/Date.jsx]: done

File: test/components/bahmni-design-system/Date.test.js (hunk 4 above)
Comment (sahupadma) [reply test/components/bahmni-design-system/Date.test.js]: done

File: src/components/bahmni-design-system/NumericBox.jsx (hunk 3 above)
Comment (sahupadma) [reply src/components/bahmni-design-system/NumericBox.jsx:31]: added

File: src/components/bahmni-design-system/NumericBox.jsx (hunk 2 above)
Comment (sahupadma) [reply src/components/bahmni-design-system/NumericBox.jsx]: done

File: src/components/bahmni-design-system/CarbonContainer.jsx (hunk 1 above)
Comment (sahupadma) [reply src/components/bahmni-design-system/CarbonContainer.jsx:15]: modified

File: src/components/bahmni-design-system/NumericBox.jsx (hunk 6 above)
Comment (sahupadma) [reply src/components/bahmni-design-system/NumericBox.jsx:143]: this is giving warnings/red borders on load of form

File: src/components/bahmni-design-system/DateTime.jsx (hunk 8 above)
Comment (sahupadma) [reply src/components/bahmni-design-system/DateTime.jsx]: done

**Review by ravinderkabli — APPROVED** [review]
Approved
